aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,379 @@
+    int error_index = 0;
+#pragma omp parallel for private(k)
+    for (k = 0; k < numChannels; k++) {
+      auto finalOutputOffset = nOutputOffset + k * outputWidth * outputHeight;
+      auto finalInputOffset = nInputOffset + k * inputWidth * inputHeight;
+      auto* output_p_k = rawOutput + finalOutputOffset;
+      auto* input_p_k = rawInput + finalInputOffset;
+      auto* ind_p_k = rawIndices + finalInputOffset;
+
+      int maxp;
+      for (auto i = 0; i < inputHeight; i++) {
+        for (auto j = 0; j < inputWidth; j++) {
+          maxp = ind_p_k[i * inputWidth + j];
+          if (maxp < 0 || maxp >= outputWidth * outputHeight) {
+#pragma omp critical
+            {
+              has_error = 1;
+              error_index = maxp;
+            }
+          } else {
+            output_p_k[maxp] = input_p_k[i * inputWidth + j];
+          }
+        }
+      }
+    }
+    if (has_error) {
+      AT_ERROR(
+          "Found an invalid max index %ld (output volumes are of size %dx%d)",
+          error_index,
+          outputHeight,
+          outputWidth);
+    }
+  }
+  return output;
+}
+
+at::Tensor& MaxUnpooling2d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+  AT_CHECK(
+      self.ndimension() == 4,
+      "Input to MaxUnpooling2d should be a NCHW Tensor");
+  AT_CHECK(
+      self.sizes() == indices.sizes(),
+      "Shape of indices should match shape of input");
+
+  // is_empty check
+  for (int64_t i = 0; i < self.ndimension(); i++) {
+    AT_CHECK(
+        self.size(i) > 0,
+        "input must be nonempty, but input has sizes: ",
+        self.sizes());
+  }
+
+  auto numBatch = self.size(0);
+  auto numChannels = self.size(1);
+  auto outputHeight = output_size[0];
+  auto outputWidth = output_size[1];
+
+  output.zero_();
+  output.resize_({numBatch, numChannels, outputHeight, outputWidth});
+
+  AT_CHECK(output.is_contiguous(), "output must be contiguous");
+
+  AT_DISPATCH_FLOATING_TYPES(
+      self.type(), "MaxUnpooling2d_forward_out_cpu_", ([&] {
+        MaxUnpooling2d_forward_out_cpu_<scalar_t>(
+            output,
+            self.contiguous(),
+            indices.contiguous(),
+            output_size[0],
+            output_size[1]);
+      }));
+  return output;
+};
+
+at::Tensor MaxUnpooling2d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      self.ndimension() == 4,
+      "Input to MaxUnpooling2d should be a NCHW Tensor",
+      self.sizes());
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+  for (int64_t i = 0; i < self.ndimension(); i++) {
+    AT_CHECK(
+        self.size(i) > 0,
+        "input must be nonempty, but input has sizes: ",
+        self.sizes());
+  }
```

Comment:
Aren't the checks here redundant because `MaxUnpooling2d_forward_out_cpu` will immediately do the same checks again? Best not to copy paste checks; they're the kind of thing where someone will most definitely forget to update them when they show up somewhere else.

Reply:
The check starting on line 121 is redundant (and will remove), but I'm not sure the first two (self.ndimension(), output_size.size() are - I explicitly added them so that when we index into `self` and `output_size` on line 127 we don't go out of bounds and get an error there.

If you're suggesting that we throw the error there in the case of bad input, I can get rid of these checks